### ISSUE TITLE
[APInt] roundToDouble: exponent off-by-one and implicit leading-1 not stripped

### ISSUE THREAD

**Issue by sunchao6106** [llvm/llvm-project#197926]
## Description

`APInt::roundToDouble()` in `llvm/lib/Support/APInt.cpp` has two bugs in its manual IEEE 754 double construction for values that don't fit in a single `uint64_t`:

### Bug 1: Exponent is off by one

```cpp
unsigned n = Tmp.getActiveBits();
uint64_t exp = n;          // BUG: should be n - 1
// ...
exp += 1023;
```

`getActiveBits()` returns `n`, meaning the leading 1 is at bit position `n-1`. The value equals `1.fraction × 2^(n-1)`, so the unbiased exponent is `n-1`, not `n`. The biased exponent should be `(n-1) + 1023 = n + 1022`, but the code produces `n + 1023`.

The overflow check `if (exp > 1023)` is correspondingly off — it should trigger at `n - 1 > 1023` (i.e. `n > 1024`), not `n > 1023`.

### Bug 2: Implicit leading 1 not stripped from mantissa

Line 961 comments "The leading bit of mantissa is implicit, so get rid of it" but the code never actually removes it:

```cpp
// The leading bit of mantissa is implicit, so get rid of it.
uint64_t sign = isNeg ? (1ULL << (APINT_BITS_PER_WORD - 1)) : 0;
uint64_t I = sign | (exp << 52) | mantissa;  // mantissa still has leading 1 at bit 51!
```

In the `hiWord == 0` path, `mantissa >>= n - 52` shifts the leading 1 to bit 51, which is then OR'd directly into the IEEE mantissa field without masking.

### Concrete example

`APInt(128, -8, /*isSigned=*/true).roundToDouble(true)`:
- Passes early return (128-bit, many active bits for negative value)
- `Tmp = 8`, `n = 4`, `hiWord = 0`
- `mantissa = 8` (leading 1 at bit 3, no shift since n ≤ 52)
- `exp = 4 + 1023 = 1027`
- Assembled double: exponent = 4, significand ≈ 1.0 → **result ≈ −16.0**
- **Correct result: −8.0** (off by exactly 2×)

### Additional issue in multi-word mantissa extraction

When `n % APINT_BITS_PER_WORD == 0` (e.g. n=128), the formula:
```cpp
uint64_t hibits = Tmp.U.pVal[hiWord] << (52 - n % APINT_BITS_PER_WORD);
```
computes `<< 52`, which shifts most significant bits out of the 64-bit word, producing garbage.

### Suggested fix

1. `exp = n - 1` (not `n`)
2. Adjust mantissa extraction to shift by `n - 53` (not `n - 52`), then mask with `(1ULL << 52) - 1` to strip the leading 1
3. Fix multi-word path to handle `n % 64 == 0` correctly

### Note

The inverse function `APIntOps::RoundDoubleToAPInt` is correct — it properly subtracts the 1023 bias and restores the implicit leading 1.

**Comment by efriedma-quic** [comment]
See #192451

**Comment by sunchao6106** [comment]
This is a duplicate of the issues addressed in #192451, which already has a fix (delegating to `APFloat::convertFromAPInt`) with CI passing. Closing in favor of that PR.
